File: services/service.py
from flask import request, jsonify, Flask
from pymongo import MongoClient
from flask_bcrypt import Bcrypt
import jwt
from datetime import datetime, timedelta, timezone
import os

app = Flask(__name__)

bcrypt = Bcrypt(app)

MONGO_URI = os.getenv("MONGO_URI")
mongo_client = MongoClient(MONGO_URI)
db = mongo_client["Anime-Galaxy"]
users_collection = db['users']
watch_later_collection = db['watch-later']


SECRET_CODE = "ANIMEGALAXY"


def register():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')
    username = data.get('username')
    
    user = {
        "username": username,
        "email": email,
        "password": password
    }
    print("User :", user)

    if not username or not password or not email:
        return jsonify({'error': 'Username, email, and password are required'}), 400
    hashed_pass = bcrypt.generate_password_hash(password).decode("utf-8")
    if users_collection.find_one({'email': email}):
        return jsonify({'error': 'User already exists'}), 400

    user_id = users_collection.insert_one({'username': username, 'email': email, 'password': hashed_pass}).inserted_id
    
    token = jwt.encode({"user": user , "exp": datetime.now(timezone.utc) + timedelta(hours=1)}, SECRET_CODE, algorithm="HS256")

    watch_later_collection.insert_one({'email': email, 'watchLaterList': []})

    return jsonify({'message': 'User registered successfully', "token": token}), 201


def login():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')

    if not email or not password:
        return jsonify({'error': 'Email and password are required'}), 400

    user = users_collection.find_one(
        {'email': email},
        {'_id': 1, "email": 1, 'username': 1, "password": 1}
    )
    
    if not bcrypt.check_password_hash(user['password'], password):
        return jsonify({'error': 'Invalid email or password'}), 401

    watch_later_data = watch_later_collection.find_one(
        {'email': email},
        {'_id': 0, 'watchLaterList': 1}
    )
    watch_later_list = watch_later_data['watchLaterList'] if watch_later_data else []

    token = jwt.encode({"user": {
        "email": email,
        "_id": str(user["_id"]),
        "username": user["username"]
        } , "exp": datetime.now(timezone.utc) + timedelta(hours=1)}, SECRET_CODE, algorithm="HS256")


    return jsonify({
        'message': 'Login successful',
        'token': token,
        'watchLaterList': watch_later_list
    }), 200
    
def store_watch_later():
    data = request.get_json()
    email = data.get('email')
    watch_later_list = data.get('watchLaterList')

    print("Received data:", data)

    if not email:
        print("Error: Email is required")
        return jsonify({'error': 'Email is required'}), 400
    if watch_later_list is None:
        print("Error: Watch Later list is required")
        return jsonify({'error': 'Watch Later list is required'}), 400
    if not isinstance(watch_later_list, list):
        print("Error: Watch Later list must be a list")
        return jsonify({'error': 'Watch Later list must be a list'}), 400

    result = watch_later_collection.update_one(
        {'email': email},
        {'$set': {'watchLaterList': watch_later_list}},
        upsert=True
    )

    print("Update result:", result.raw_result)
    return jsonify({'message': 'Watch Later list updated successfully'}), 200

def change_password():
    data = request.get_json()
    email = data.get('email')
    # oldPass = data.get("oldPassword")
    new_password = data.get('newPassword')

    if not email or not new_password:
        return jsonify({'error': 'Email and new password are required'}), 400

    user = users_collection.find_one({'email': email})
    if not user:
        return jsonify({'error': 'User not found'}), 404

    users_collection.update_one({'email': email}, {'$set': {'password': new_password}})
    return jsonify({'message': 'Password updated successfully'}), 200

File: services/service.js
const jwt = require("jsonwebtoken");
const getDB = require("../db");

const login = async (req, res) => {
    try {
        const db = await getDB.getDatabase();
        const User_collection = db.collection("users");
        const WatchLater_collection = db.collection("watch-later");

        const email = req.body.email;
        const password = req.body.password;

        const user = await User_collection.findOne({'email': email, 'password': password});
        if(!user){
            return res.status(404).json({status: "error", message: "User not found"});
        }

        console.log(user);
        const token  = jwt.sign({"email": email, "username": user.username},"ANIMEGALAXY", {expiresIn:"5h"});

        const watchLater = await WatchLater_collection.findOne({"email": email}, {"_id": 0 ,"watchLaterList": 1});

        return res.status(200).json({"message": "Login successful", "token": token, "watchLaterList": watchLater});
        
    } catch (error) {
        console.log("error :", error);
        return res.status(500).json({"message": error.message});
    }
};

const register = async (req, res) => {
    try {
        const db = await getDB.getDatabase();
        const User_collection = db.collection("users");
        const WatchLater_collection = db.collection("watch-later");

        const email = req.body.email;
        const username = req.body.username;
        const password = req.body.password;

        const findUser = await User_collection.findOne({"email": email});
        if(findUser){
            return res.status(409).json({status: "error", message: "User already exists"});
        }

        const InsertUser = await User_collection.insertOne({"email": email, "username": username, "password": password});
        if(!InsertUser){
            return res.status(500).json({status: "error", message: "User not created"});
        }
        const InsertUserWatchLater = await WatchLater_collection.insertOne({"email": email, "watchLaterList": []});
        const token = jwt.sign({"email": email, "username": username},"ANIMEGALAXY", {expiresIn:"5h"});
        return res.status(200).json({"message": "User created successfully", "token": token});

    } catch (error) {
        console.log("error :", error);
        return res.status(500).json({"message": error.message});
    }
}

const changePassword = async (req, res) => {
    try {
        const db = await getDB.getDatabase();
        const User_collection = db.collection("users");

        const email = req.body.email;
        const newPassword = req.body.newPassword;

        const updatepass = await User_collection.updateOne({"email": email},{"$set": {"password": newPassword}});
        if(!updatepass){
            return res.status(500).json({status: "error", message: "Password not updated"});
        }
        return res.status(200).json({"message": "Password updated successfully"});
    } catch (error) {
        console.log("error :", error);
        return res.status(500).json({"message": error.message});
    }
}

const store_watch_later = async (req, res) => {
    try {
        const db = await getDB.getDatabase();
        const WatchLater_collection = db.collection("watch-later");
        const email = req.body.email;
        const password = req.body.password;
        const WatchLaterList = req.body.watchLaterList;
        const UserFind = await findUser(email, password);
        if(!UserFind){
            return res.status(404).json({"message": "User not found"});
        }

        const UpdateWatchLater = await WatchLater_collection.updateOne({"email": email},{"$set": {"watchLaterList": WatchLaterList}},{upsert: true});
        if(!UpdateWatchLater){
            return res.status(409).json({"message": "Watch Later list not updated"});
        }
        return res.status(200).json({"message": "Watch Later list updated successfully"});
    } catch (error) {
        console.log("error :", error);
        return res.status(500).json({"message": error.message});
    }
}

async function findUser(email, password) {
    const db = await getDB.getDatabase();
    const User_collection = db.collection("users");
    const Find = await User_collection.findOne({ "email": email, "password": password });
    if (!Find) {
        return false;
    }
    return true;
}

module.exports = {login, register, changePassword, store_watch_later};